internal/v1alpha2/source/image.go

```diff
@@ -0,0 +1,315 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package source
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"path/filepath"
+	"strings"
+
+	"github.com/operator-framework/rukpak/api/v1alpha2"
+	"github.com/operator-framework/rukpak/internal/util"
+	"github.com/operator-framework/rukpak/internal/v1alpha2/store"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	applyconfigurationcorev1 "k8s.io/client-go/applyconfigurations/core/v1"
+	v1 "k8s.io/client-go/applyconfigurations/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// Image implements unpack interface.
+type image struct {
+	Client       client.Client
+	KubeClient   kubernetes.Interface
+	PodNamespace string
+	UnpackImage  string
+}
+
+const imageUnpackContainerName = "bundle"
+
+type logfn func(ctx context.Context, pod *corev1.Pod) ([]byte, error)
+
+func (i *image) Unpack(ctx context.Context, source *v1alpha2.BundleDeplopymentSource, store store.Store, opts UnpackOption) (*Result, error) {
+	// Validate inputs
+	if err := i.validate(source, opts); err != nil {
+		return nil, fmt.Errorf("validation unsuccessful during unpacking %v", err)
+	}
+	// storage path to store contents in local directory.
+	storagePath := filepath.Join(store.GetBundleDeploymentName(), filepath.Clean(source.Destination))
+	return i.unpack(ctx, storagePath, source, store, opts)
+}
+
+func (i *image) unpack(ctx context.Context, storagePath string, source *v1alpha2.BundleDeplopymentSource, store store.Store, opts UnpackOption) (*Result, error) {
+	pod := &corev1.Pod{}
+	op, err := i.ensureUnpackPod(ctx, store.GetBundleDeploymentName(), *source, pod, opts)
+	if err != nil {
+		return nil, err
+	} else if op == controllerutil.OperationResultCreated || op == controllerutil.OperationResultUpdated || pod.DeletionTimestamp != nil {
```

Comment:
If `i.ensureUnpackPod()` did not return Created or Updated, why is a deleting pod a pending unpack?